Fixed #25759 -- Add **kwargs to as_sql method of Expressions

kaifeldhoff opened the pull request:
https://code.djangoproject.com/ticket/25759

jarshwah commented:
For your first question, I'd consider doing something like this:

```
data = self.extra.copy()
data.update(**kwargs)
template = data.get('template', self.template)
data['expressions'] = data['field'] = self.arg_joiner.join(sql_parts)
return template % data, params
```

That's off the top of my head, let me know if you have issues and I'll advise.

Your part 2 is correct.

Make those changes and I'll give a thorough review. Good work!

kaifeldhoff commented:
Thanks  jarshwah,

where I'm not sure:
The current django-code manipulates `self.extra`. Your code does not. Is this OK?

I'll go ahead like this and see what the tests show.

Kai

jarshwah commented:
Yes, manipulating `self.extra` is wrong. That code (which I wrote) is buggy if used on multiple databases with different backends. Testing for that use case is difficult though, because the django tests don't run different backends on multiple databases.

jarshwah commented on django/db/models/functions.py:
```diff
@@ -102,9 +102,12 @@ def __init__(self, *expressions, **extra):
-    def as_sqlite(self, compiler, connection):
+    def as_sqlite(self, compiler, connection, **kwargs):
         """Use the MAX function on SQLite."""
-        return super(Greatest, self).as_sql(compiler, connection, function='MAX')
```
This line doesn't need to change. Passing `function='Max'` as a kwarg will do the right thing, since the super implementation will do the mixing of `self.extra` and `**kwargs` for you.

jarshwah commented on django/db/models/functions.py:
```diff
@@ -122,9 +125,12 @@ def __init__(self, *expressions, **extra):
-    def as_sqlite(self, compiler, connection):
+    def as_sqlite(self, compiler, connection, **kwargs):
         """Use the MIN function on SQLite."""
-        return super(Least, self).as_sql(compiler, connection, function='MIN')
```
As above. This line doesn't need to change.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -224,13 +224,13 @@ def process_rhs(self, compiler, connection):
         if self.rhs_is_direct_value() and max_in_list_size and len(self.rhs) > max_in_list_size:
-            return self.split_parameter_list_as_sql(compiler, connection)
-        return super(In, self).as_sql(compiler, connection)
+            return self.split_parameter_list_as_sql(compiler, connection, **kwargs)
```
I'm not sure that this particular call requires `**kwargs`. Check the actual function definition to see if it will require kwargs or not. I don't think it is overridable with `as_vendorname`.

jarshwah commented on django/db/models/lookups.py:
```diff
@@ -224,13 +224,13 @@ def process_rhs(self, compiler, connection):
+        return super(In, self).as_sql(compiler, connection, **kwargs)
 
-    def split_parameter_list_as_sql(self, compiler, connection):
+    def split_parameter_list_as_sql(self, compiler, connection, **kwargs):
```
I don't think this method requires `**kwargs` because it's not part of the expressions API, it just looks similar.

kaifeldhoff commented:
I added my work.

> I'm not sure that this particular call requires **kwargs. Check the actual function definition to see if it will require kwargs or not. I don't think it is overridable with as_vendorname.

I don't seen any use of extra/kwargs in super-function, but I have no idea, if it is overridable.

kaifeldhoff commented:
Standard tests pass.

jarshwah commented on django/db/models/expressions.py:
```diff
@@ -834,11 +833,11 @@ def copy(self):
         if not self.cases:
             return compiler.compile(self.default)
-        template_params = dict(extra) if extra else {}
+        template_params = kwargs.get('extra', {})
```
This should just be `template_params = kwargs`. The `extra` in the signature that was removed is acting like a `kwargs` anyway.

kaifeldhoff commented:
OK, I just wanted to ensure the previous default behaviour.

jarshwah commented on django/db/models/functions.py:
```diff
@@ -27,7 +27,7 @@ class ToNCLOB(Func):
             expressions = [
                 ToNCLOB(expression) for expression in self.get_source_expressions()]
             self.set_source_expressions(expressions)
```
What this should do here is something like:

```
c = self.copy()
c.set_source_expressions(expressions)
return super(Coalesce, c).as_sql(compiler, connection, **kwargs)
```

The theme here is to stop mutating `self` within `as_sql` or `as_postgres/oracle` etc, which is the major point of adding `kwargs`.

kaifeldhoff replied:
Check this in next commit. Hope I got it right.

jarshwah commented on django/db/models/functions.py:
```diff
@@ -41,17 +41,17 @@ class ConcatPair(Func):
         self.function = 'CONCAT_WS'
         self.template = "%(function)s('', %(expressions)s)"
-        return super(ConcatPair, self).as_sql(compiler, connection)
+        return super(ConcatPair, self).as_sql(compiler, connection, **kwargs)
```
Remove `self.function = 'CONCAT_WS' and the following line which mutates `self.template` and instead:

```
return super(ConcatPair, self).as_sql(compiler, connection, function='CONCAT_WS', template="%(function)s('' %(expression)s)"
```

jarshwah commented on django/db/models/functions.py:
```diff
@@ -136,9 +136,9 @@ def __init__(self, expression, **extra):
+    def as_mysql(self, compiler, connection, **kwargs):
         self.function = 'CHAR_LENGTH'
-        return super(Length, self).as_sql(compiler, connection)
+        return super(Length, self).as_sql(compiler, connection, **kwargs)
```
Remove `self.function = ..` and move it to a kwarg of the `as_sql` call below:

```
return super(Length, self).as_sql(compiler, connection, function='CHAR_LENGTH')
```

jarshwah commented on django/db/models/functions.py:
```diff
@@ -154,12 +154,12 @@ def __init__(self, output_field=None, **extra):
         # other databases.
         self.template = 'STATEMENT_TIMESTAMP()'
-        return self.as_sql(compiler, connection)
+        return self.as_sql(compiler, connection, **kwargs)
```
Remove `self.template = ..` and replace with:

```
return self.as_sql(compiler, connection, template='STATEMENT_TIMESTAMP()')
```

jarshwah commented on django/db/models/functions.py:
```diff
@@ -182,13 +182,13 @@ def __init__(self, expression, pos, length=None, **extra):
+    def as_oracle(self, compiler, connection, **kwargs):
         self.function = 'SUBSTR'
-        return super(Substr, self).as_sql(compiler, connection)
+        return super(Substr, self).as_sql(compiler, connection, **kwargs)
```
This and one above, replace `self.function = ..` with `function=...` in as_sql method call.

jarshwah commented:
Code looks good, a few things to clean up now that kwargs exists everywhere. Next step will be documenting the change in release notes, and adding the `**kwargs` signature to the expressions API doc. The ticket also makes mention of editing the part in the docs that shows overriding `as_sql` to recommend using the kwargs rather than mutating `self.template` or `self.function`.

There are also probably some tests that should be rewritten in the new way of using `kwargs`. Mainly the expressions and annotations test packages.

timgraham commented:
buildbot, add to whitelist.

charettes commented on django/db/models/expressions.py:
```diff
@@ -527,13 +527,12 @@ def as_sql(self, compiler, connection, function=None, template=None):
-        return template % self.extra, params
+        data = self.extra.copy()
+        data.update(**kwargs)
+        data['function'] = data.get('function', self.function)
```
`data.setdefault('function', self.function)`

charettes commented on django/db/models/expressions.py:
```diff
@@ -527,13 +527,12 @@ def as_sql(self, compiler, connection, function=None, template=None):
+        data = self.extra.copy()
+        data.update(**kwargs)
+        data['function'] = data.get('function', self.function)
+        template = data.get('template', self.template)
```
`data.setdefault('template', self.template)`

kaifeldhoff commented:
I will go on. Hope the code becomes clearer for me while working on the tests.

kaifeldhoff commented:
Looked into annotations- and expressions-tests and  found only this case, where the new signature can be used. Is this what you mean? I found classes overriding as_sql in custom-lookups-tests. May these be transfered into the formulation defined in clss Func(Expression)?

To be sure I got the clue: The worst case reason for this ticket is a lockup/transform-class with two or more as_vendor-definitions, all using  some class- or instance-variables. When overriding only one as_vendor-definition, these variables may be modified, what may lead into a broken behavior in some other as_vendor-definition. Is this the case you had in mind when creating the ticket? If not, please explain the worst case, then I may understand the solution.

kaifeldhoff commented:
Seems I didn't get the clue. Cannot make the tests pass.

jarshwah commented on django/db/models/functions.py:
```diff
@@ -41,17 +42,17 @@ class ConcatPair(Func):
-        self.template = "%(function)s('', %(expressions)s)"
-        return super(ConcatPair, self).as_sql(compiler, connection)
+        return super(ConcatPair, self).as_sql(
+            compiler, connection, function='CONCAT_WS', template="%(function)s('', %(expression)s)")
```
`%(expressions)s)` not `%(expression)s)`. You're missing the `s` at the end of `expressions`

jarshwah commented:
There's also a few cases where you've supplied a keyword argument as `function = 'function_name'`. You should remove the spaces either side of the `=` such that `, function='function_name'`.

I'm not sure why one of the tests is failing, but maybe it'll resolve itself with the other tests fixed.

kaifeldhoff commented:
Hi,

Is there any more work to be done for the tests?

Thanks

Kai

kaifeldhoff commented:
Hi,

got the main work done. Awaiting your review.

Points to remark:
* Doc for 3rd-party backends: I'm not sure if I got the right source here or if there are open points left form ticket #25687
* I didn't find much to change in the test, neither did I wrote new tests. Any feedback would be appreciated.

Thanks

Kai

kaifeldhoff commented:
After reading the docs how to fix the docs- and flake8-errors from my point of view the work is done.

The recommendation in the expressions-doc to use kwargs and not self-variables is moved into a note nearby the example for custom expressions. This makes more sense for me than in the section for third-party backends.

Happy reviewing!

Kai

timgraham commented on django/db/models/functions.py:
```diff
@@ -17,17 +17,18 @@ def __init__(self, *expressions, **extra):
+    def as_oracle(self, compiler, connection, **kwargs):
         # we can't mix TextField (NCLOB) and CharField (NVARCHAR), so convert
         # all fields to NCLOB when we expect NCLOB
+        c = self.copy()
```
better variable name?

jarshwah replied:
It's mostly consistent with other copies in the same file, but `clone = ` and `copy = ` is also used. I think it's probably better to use `copy`.

timgraham replied:
I guess the only disadvantage is the possibility to shadow the Python's built-in `copy`. If it's used elsewhere in the file, I'm okay with it.

timgraham commented on django/db/models/functions.py:
```diff
@@ -41,17 +42,16 @@ class ConcatPair(Func):
-        self.template = "%(function)s('', %(expressions)s)"
-        return super(ConcatPair, self).as_sql(compiler, connection)
+        return super(ConcatPair, self).as_sql(
+            compiler, connection, function='CONCAT_WS', template="%(function)s('', %(expressions)s)")
```
put the closing parenthesis on the next line

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,38 @@ expressions::
-        self.template = 'coalesce( %(expressions)s )'
-        return self.as_sql(compiler, connection)
+        return self.as_sql(compiler, connection,
+            template='coalesce( %(expressions)s )'
```
single should be fine here I think, what you have currently violates a flake8 warning I think

kaifeldhoff replied:
As single line this would exceed 79 columns. Or what do you mean?

timgraham replied:
Yes, for code samples the 79 character limit isn't important.

kaifeldhoff replied:
OK, Done.

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,38 @@ expressions::
+
+.. note::
+
+    When working with different database-backends and multiple databases you'll
```
I don't know if the note is necessary. I don't think anyone would use the alternate technique instead of following the documentation? We should have a ``.. versionchanged:: 1.10` note that describes how the behavior changed. You could mention the reason for that change there.

timgraham commented on django/db/models/functions.py:
```diff
@@ -182,13 +180,11 @@ def __init__(self, expression, pos, length=None, **extra):
-    def as_sqlite(self, compiler, connection):
-        self.function = 'SUBSTR'
-        return super(Substr, self).as_sql(compiler, connection)
+    def as_sqlite(self, compiler, connection, **kwargs):
```
I don't understand the reason for changes like this one. Why does the method need to accept unused `**kwargs`?

jarshwah replied:
Agreed. Either pass `**kwargs` through to the underlying `as_sql` method, or remove them from the signature of `as_sqlite`. I'd lean towards passing kwargs through to remain consistent with the API. I can't think of any reason why the `as_sqlite` would ever need to accept `kwargs` (since it is the one doing the overriding) though. Tim?

timgraham replied:
Removing `**kwargs` from the `as_vendor()` methods sounds good to me.

timgraham commented on docs/ref/models/lookups.txt:
```diff
@@ -83,7 +83,7 @@ field references, aggregates, and ``Transform`` are examples that follow this
 following methods:
 
-.. method:: as_sql(self, compiler, connection)
+.. method:: as_sql(self, compiler, connection, **kwargs)
```
Need to explain `**kwargs`.

jarshwah replied:
And call out that `**kwargs` will be mixed with `self.extra` for the purposes of interpolating into `self.template`. I'd also callout that `template` kwarg is a special case which can be used to override `self.template` for that particular invocation.

timgraham commented on django/db/models/expressions.py:
```diff
@@ -519,7 +519,7 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
         return c
 
-    def as_sql(self, compiler, connection, function=None, template=None):
+    def as_sql(self, compiler, connection, **kwargs):
```
Is it best to remove common parameters like `function` and `template`? Seems to me it improves readability a bit to have common kwargs explicitly declared.

jarshwah replied:
Probably not a bad idea, but it does make the implementation slightly more complicated.

timgraham replied:
This comment isn't addressed -- any conclusions about it?

jarshwah replied:
Ok, let's change this to:

```
def as_sql(self, compiler, connection, function=None, template=None, arg_joiner=None, **extra_context):
```

You'll need to conditionally add function to extra_context before merging it with self.extra:

```
if function is not None:
    extra_context['function'] = function
arg_joiner = arg_joiner or self.arg_joiner
template = template or self.template
```

etcetera.

The docs for `Func` will need to call out the extra key word arguments for as_sql specifically.

kaifeldhoff commented on django/db/models/expressions.py:
```diff
@@ -300,9 +300,9 @@ def relabeled_clone(self, change_map):
+        clone = copy.copy(self)
+        clone.copied = True
+        return clone
 
```
Here ``copy`` cannot be used as variable-name, so I decided to use ``clone``.

kaifeldhoff commented on django/db/models/functions.py:
```diff
@@ -44,23 +45,22 @@ def __init__(self, left, right, **extra):
-        coalesced.template = '%(expressions)s'
-        return super(ConcatPair, coalesced).as_sql(compiler, connection)
+        return super(ConcatPair, coalesced).as_sql(compiler, connection, template='%(expressions)s')
 
```
This tells me that the ``arg_joiner`` is also backend-dependant. Should we also use this as kwarg in ``Func``-class?

charettes replied:
I think it would make sense.

kaifeldhoff commented:
Hm, when I update my local master an merge in my local branch I get no conflicts. So I have no hint, where the conflict is.

Any ideas?

charettes commented:
You original commit is based on 7424ad07748c34eba1aae346213e532e64586def, what you want to here is _rebase_ your `ticket_25759` branch on the current master branch. You might want to squash your commits at the same time. Here's a detailed article about rebasing: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

kaifeldhoff commented:
Finally it works! Thanks!

jarshwah commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,38 @@ expressions::
             sql, params = compiler.compile(expression)
             sql_expressions.append(sql)
             sql_params.extend(params)
         self.extra['expressions'] = ','.join(sql_expressions)
```
I would add the `data = kwargs.copy()` way of implementing `as_sql`, very similar to the actual implementation of `Func.as_sql` unless you think that'd confuse the implementation? This method here is mutating `self.extra` which we're trying to advocate against.

jarshwah commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,38 @@ expressions::
+.. versionchanged:: 1.10
+
+    ``as_sql``-methods now support ``kwargs`` to override the default
+    ``template``- or ``function``-variables of the ``Expression``-class. This
```
`function` isn't an attribute of `Expression`, it's an attribute of `Func` which makes this sentence slightly wrong in this context. I would write it like this:

" ``as_sql``-methods now support ``kwargs`` to override the default ``template`` and to provide extra context to be used when interpolating into the ``template``."

jarshwah commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,38 @@ expressions::
+    is useful for 3rd-party database-backends that may have different
+    SQL-syntax. Simply override ``as_sql`` with your backend-specific
+    ``as_vendorname``-method and call the existing ``as_sql``-method with the
+    modified ``template``- or ``function``-parameter.
```
As above, rewrite as:

... "with the modified ``template`` or context parameters."

jarshwah commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,38 @@ expressions::
+    When working with different database-backends and multiple databases the
+    new procedure is needed, because all backends share the same ``Expression``
+    -class. So mutating the class-variables for ``template`` or ``function``
+    will affect all backends which is usually not wanted.
```
I would change this paragraph to something like:

"Since expressions can be shared between queries it's important to never mutate state like ``self.template`` or ``self.extra``. The use of ``**kwargs`` is intended to override behaviour of ``as_sql`` for that particular call, without affecting future calls to ``as_sql``."

jarshwah commented on docs/ref/models/lookups.txt:
```diff
@@ -94,6 +94,13 @@ following methods:
+    may have different SQL-syntax. Simply override ``as_sql`` with your
+    backend-specific ``as_vendorname``-method and call the existing ``as_sql``-
+    method with the modified ``template``- or ``function``-parameter.
+
```
This particular paragraph is tricky because it's referring to the Query Expression API and not to Expressions more generally (which Lookups are not). Here's my attempt at writing it:

"
``kwargs`` may be used to supply extra context to the ``as_sql`` method for that particular call. This can be useful for Expressions in particular, where ``template`` can be overridden and variables intended for interpolating the ``template`` string can be passed. 3rd-party backends may have different SQL-syntax so they can pass these differences through via ``kwargs``.

Example: ``self.as_sql(compiler, connection, template="%(function)s(%(arg)s)", function='Print', arg='Hello World')``
"

jarshwah commented on docs/releases/1.10.txt:
```diff
@@ -202,6 +202,9 @@ Database backends
 * Temporal data subtraction was unified on all backends.
+* Added ``**kwargs`` to ``as_sql`` method of Expressions to override behavior
+  through arguments rather than mutating ``self`` itself, which is not safe when using
+  multiple databases with different backends.
```
... " rather than mutating ``self``, which is not safe when using multiple databases with different backends."

kaifeldhoff commented on docs/ref/models/lookups.txt:
```diff
@@ -94,6 +94,19 @@ following methods:
+
+        self.as_sql(compiler, connection,
+            template="%(function)s(%(arg)s)", function='Print', arg='Hello World'
+        )
```
``arg`` is right now not supported. Shall I add this? I'm not sure if this kind of usage will be used in real-world tasks.

jarshwah replied:
``arg`` in this example is supported because it exists as a placeholder within the ``template`` that we're also passing through. This example was designed to highlight that ``kwargs`` can be interpolated into the ``template`` string.

@timgraham what do you think about the wording and example here? Would a more realistic example be better?

timgraham replied:
It's not obvious to me where this `self.as_sql()` call lives. Could you try making the example a bit more complete for a newbie reader, similar to https://docs.djangoproject.com/en/dev/ref/models/expressions/#adding-support-in-third-party-database-backends?

jarshwah replied:
How's this? I'm struggling to find an example that would add it's own private interpolation variables within the template and then supply them as a kwarg rather than just embedding the data within the template to begin with. For that reason I'm not so sure that it should be highlighted here as a main feature anymore.

```
Example::

# Concat function implementation for sqlite
def as_sqlite(self, compiler, connection):
    return self.as_sql(compiler, connection, template='%(expressions)s', arg_joiner=' || ')
```

timgraham replied:
function -> method, correct? The actual code in django/db/models/functions.py is a bit different -- is the difference intentional? Maybe something like this could be nice:
```python
.. snippet::
    :filename: django/db/models/functions.py

    class ConcatPair(Func):
        ...
        def as_sqlite(self, compiler, connection):
            coalesced = self.coalesce()
            return super(ConcatPair, coalesced).as_sql(
                compiler, connection, template='%(expressions)s', arg_joiner=' || '
            )
```

kaifeldhoff replied:
I'm not familiar enough with SQL to help construct an appropriate example.

But the more I'm listening to you both, the more I'm learning!

jarshwah replied:
(Sorry I took so long to respond, I didn't notice Tim's response until today).

The difference was mostly intentional. Showing `self.coalesce()` here isn't relevant to the example we're highlighting, and is confusing without prior context (ensuring all arguments are non-null). The ConcatPair name is an implementation detail which isn't relevant to the example either.

That said, my example still lacks context by excluding the class definition. Also, if we're going to link to an actual example in code, it's probably important that the code in the docs matches the code we're linking to.

Happy to go with Tim's example here.

kaifeldhoff commented:
I'm impressed how much work is done one the branch. Just took a small break and voila: Merge conflicts.

I've added Tim's proposal as suggested and added some notes for better understanding. Also I added the second ``as_vendor``-method, as it shows the same concept in a bit different way.

Just give me a note if more changes are needed.

timgraham commented on django/db/models/expressions.py:
```diff
@@ -202,13 +202,13 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
-        c = self.copy()
-        c.is_summary = summarize
-        c.set_source_expressions([
+        clone = self.copy()
```
It would be better to make these renames in a separate commit since it's an unrelated cosmetic cleanup.

timgraham commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,40 @@ expressions::
+
+.. versionchanged:: 1.10
+
+    ``as_sql``-methods now support ``kwargs`` to override the default
```
Try to make the note "self-contained" as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features.

kaifeldhoff commented:
I undid all the cosmetic changes. You're right there, reviewing is much easier now which is important as the changes cannot be explicitly tested.

Regarding “self-contained”-versionchange the docs aren't very clear to me. Please review and give me a note, in case I didn't get the clue.

kaifeldhoff commented:
My diff view is a bit different here regarding the Hm, my local diff lookes a bit different regarding the c/copy/clone-changes. I will check that now...

jarshwah commented on docs/ref/models/expressions.txt:
```diff
@@ -588,22 +588,40 @@ expressions::
+context parameters.
+
+Since expressions can be shared between queries it's important to never mutate
+states like ``self.template`` or ``self.extra``. The use of ``*kwargs`` is
```
"states" -> "state"
"*kwargs" -> "**kwargs"

kaifeldhoff replied:
I changed all entries to ``**kwargs``. Is this what you mean?

jarshwah commented on docs/ref/models/lookups.txt:
```diff
@@ -94,6 +94,38 @@ following methods:
+    backends may have different SQL-syntax so they can pass these differences
+    through via ``kwargs``.
+
+    The following example is taken out of the django backends modules. It
```
"The following example is taken from the Django ``ConcatPair`` expression.

kaifeldhoff replied:
Done

jarshwah commented on docs/ref/models/lookups.txt:
```diff
@@ -94,6 +94,38 @@ following methods:
+
+    The following example is taken out of the django backends modules. It
+    shows the use of ``kwargs`` for overriding ``as_sql``-methods. Note that
+    the class do not need an ``as_sql``-method. By providing the class-variable
```
"the class does not need"

jarshwah commented on docs/ref/models/lookups.txt:
```diff
@@ -94,6 +94,38 @@ following methods:
+
+    The following example is taken from the Django ConcatPair-expression.. It
+    shows the use of ``**kwargs`` for overriding ``as_sql``-methods. Note that
+    the class do not need an ``as_sql``-method. By providing the class-variable
```
Still not fixed. "the class do not need an" "the class does not need". Change the word "do" to "does".

kaifeldhoff commented:
Everything OK now?

jarshwah commented:
LGTM. I'd like one other person to give the docs another look over though. Perhaps if @timgraham gets a chance?

timgraham commented on django/db/models/expressions.py:
```diff
@@ -456,7 +456,7 @@ class TemporalSubtraction(CombinedExpression):
         super(TemporalSubtraction, self).__init__(lhs, self.SUB, rhs, output_field=fields.DurationField())
 
-    def as_sql(self, compiler, connection):
+    def as_sql(self, compiler, connection, **kwargs):
```
I'm still uncertain about the purpose or value in adding unused `**kwargs` in places like this?

timgraham commented on django/db/models/expressions.py:
```diff
@@ -995,14 +995,15 @@ def set_source_expressions(self, exprs):
         return [self.expression]
 
-    def as_sql(self, compiler, connection):
+    def as_sql(self, compiler, connection, **kwargs):
```
What's the value in accepting `**kwargs` as opposed the the specific values that are used (`template`)?

jarshwah replied:
This method should actually mix `placeholders` with `**kwargs`, which expands the usefulness of `**kwargs` to the `ordering` interpolation variable, but also any other interpolation variables that might be introduced in a subclass.

timgraham replied:
What about an `extra_placeholders` parameter (which could be a dict) instead of `kwargs`? To me, `kwargs` isn't a descriptive name and makes the code difficult to understand. Alternatively, you could rename `**kwargs` to `**extra_placeholders` and declare other parameters like `template` as explicit kwargs?

jarshwah replied:
Yeah I think that's a good idea in general. I was thinking that a general `**kwargs` added to the signature in all cases was probably correct, but I've come around to your thinking. Most uses outside of `Func` expressions just ignore the `**kwargs` parameter and create noise.

So this specific call should be (assuming "extra_context" is a good name):

```
def as_sql(self, compiler, connection, template=None, ordering=None, **extra_context):
    if template is not None:
        extra_context['template'] = template
    if ordering is not None:
        extra_context['ordering'] = ordering
    extra_context['ordering'] = ordering
    place_holders = { .. }
    place_holders.update(extra_context)
   ...
```

timgraham commented on django/db/models/lookups.py:
```diff
@@ -111,7 +111,7 @@ def get_group_by_cols(self):
         return cols
 
-    def as_sql(self, compiler, connection):
+    def as_sql(self, compiler, connection, **kwargs):
```
Again, not sure if all the unused `**kwargs` additions in this file accomplishes anything practical?

jarshwah replied:
Agreed. Let's ditch `**kwargs` in lookups if they aren't being used for anything.

kaifeldhoff commented on django/db/models/lookups.py:
```diff
@@ -362,7 +362,7 @@ def process_rhs(self, compiler, connection):
 
-    def as_sql(self, compiler, connection):
+    def as_sql(self, compiler, connection, **kwargs):
         sql, params = compiler.compile(self.lhs)
```
At this position even ``connection`` is not used either. But it remains there because it is the ``as_sql``-function signature. I think this counts also for ``**kwargs``.

The benefit of ``**kwargs`` is also where super-methods are called and use ``**kwargs``. If in the future the super-method makes use of more kwargs, then the child-methods do not have to be adapted. Just a thought, maybe I'm wrong here.

The other consistent choice would be to remove ``**kwargs`` where it is not used in the method or through a call to a super-method that makes use of it.
Even further this could be changed from ``**kwargs`` to the single kwargs like ``template``, ``function`` or ``arg_joiner``.
This would not change the purpose of this ticket: Do not touch ``self`` in ``as_vendor``.

I'm OK with the new function-signature but will do any further changes if needed.

jarshwah replied:
`template` `function` and `arg_joiner` are only common to `Func` expressions. But `Func` expressions also allow arbitrary kwargs as input into `template` interpolation. Making this change in the general case is not right. I'll comment specifically on other entries.

jarshwah commented:
Ok, sorry to do this to you @kaifeldhoff but I think Tim is right. Let's only add ``**kwargs`` to expressions that actually use kwargs or call superclasses that use kwargs (like Func and all derived Funcs/Transforms). `OrderBy` is another one that can accept kwargs.

Additionally, let's rename `**kwargs` to `extra_context` if the purpose is to interpolate into `template`. Any common key word arguments should be defined explicitly.

Classes that now accept key word arguments should have those documented in specific as_sql methods.

If this is too much to do, let me know, and I can finish off the patch for you.

kaifeldhoff commented:
This ticket is far away from being straight-forward. So I'm mostly doing what you two tell me. Additionally other tickets depend on this work and are pending.

It is far more efficient, when you take over now. It's OK for me. I'll proceed with another easy-picking.

jarshwah commented:
Sorry, it's my fault. The direction and initial instructions weren't very clear. I'll finish it up but leave the author as you since you've done the bulk of the work. Thanks for keeping up and following through though, you did a good job.

jarshwah commented:
See #6302
